testng-core/src/main/java/org/testng/internal/BaseTestMethod.java

```diff
@@ -817,7 +818,11 @@ private IRetryAnalyzer getRetryAnalyzerConsideringMethodParameters(ITestResult t
       this.m_retryAnalyzer = computeRetryAnalyzerInstanceToUse(tr);
       return this.m_retryAnalyzer;
     }
-    final String keyAsString = getSimpleName() + "#" + getParameterInvocationCount();
+
+    final String keyAsString =
+        getSimpleName()
+            + "#"
+            + (isDataDriven() ? hashParameters(tr) : getParameterInvocationCount());
```

Comment:
Why not both in all cases?

Reply:
Yeah that could be done too. I was just trying to stick to, handling the bug 😄

Reply:
I will make the change and push it in. Please let me know if there is anything else that needs to be changed

Reply:
This is now fixed. Please check @juherr